src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs

```diff
@@ -0,0 +1,118 @@
+        {
+            if (document == null)
+            {
+                return ImmutableArray<CodeAction>.Empty;
+            }
+
+            var (codeFixCollections, codeRefactorings) = await GetCodeFixesAndRefactoringsAsync(
+                document, codeFixService,
+                codeRefactoringService, selection,
+                cancellationToken).ConfigureAwait(false);
+
+            var codeActions = codeFixCollections.SelectMany(c => c.Fixes.Select(f => f.Action)).Concat(
+                    codeRefactorings.SelectMany(r => r.CodeActions.Select(ca => ca.action)));
+
+            return codeActions;
+        }
+
+        public static async Task<(ImmutableArray<CodeFixCollection>, ImmutableArray<CodeRefactoring>)> GetCodeFixesAndRefactoringsAsync(
+            Document document,
+            ICodeFixService codeFixService,
+            ICodeRefactoringService codeRefactoringService,
+            LSP.Range selection,
+            CancellationToken cancellationToken)
+        {
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var textSpan = ProtocolConversions.RangeToTextSpan(selection, text);
+
+            var codeFixCollections = Array.Empty<CodeFixCollection>().ToImmutableArray();
+            var codeRefactorings = Array.Empty<CodeRefactoring>().ToImmutableArray();
+
+            var codeFixCollectionsTask = Task.Run(
+                async () => codeFixCollections = await codeFixService.GetFixesAsync(
+                    document, textSpan, includeSuppressionFixes: true, cancellationToken).ConfigureAwait(false));
+            var codeRefactoringsTask = Task.Run(
+                async () => codeRefactorings = await codeRefactoringService.GetRefactoringsAsync(
+                    document, textSpan, cancellationToken).ConfigureAwait(false));
+
+            await Task.WhenAll(codeFixCollectionsTask, codeRefactoringsTask).ConfigureAwait(false);
+            return (codeFixCollections, codeRefactorings);
+        }
+
+        public static CodeAction? GetCodeActionToResolve(string distinctTitle, ImmutableArray<CodeAction> codeActions)
+        {
+            // Searching for the matching code action. We compare against the unique identifier
+            // (e.g. "Suppress or Configure issues.Configure IDExxxx.Warning") instead of the
+            // code action's title (e.g. "Warning") since there's a chance that multiple code
+            // actions may have the same title (e.g. there could be multiple code actions with
+            // the title "Warning" that appear in the code action menu if there are multiple
+            // diagnostics on the same line).
+            foreach (var c in codeActions)
+            {
+                var action = CheckForMatchingAction(c, distinctTitle, currentTitle: "");
+                if (action != null)
+                {
+                    return action;
+                }
+            }
+
+            return null;
```

Comment:
is this possible?  doesit in indicate a bug somewhere?  should we be assert/logging instead?

Reply:
in other words, code like this scares me.  it makes me feel lke we'll have a customer saying "my code actions aren't behaving properly" and we'll never be able to track down why.

Reply:
I think you're right. I pushed a commit that modifies all callers of the method to catch this case.